lib/cli/src/commands/publish.rs

```diff
@@ -0,0 +1,652 @@
+        })?;
+    }
+
+    Ok((readme, license_file))
+}
+
+fn append_path_to_tar_gz(
+    builder: &mut tar::Builder<Vec<u8>>,
+    base_path: &Path,
+    target_path: &Path,
+) -> Result<PathBuf, (PathBuf, io::Error)> {
+    let normalized_path = normalize_path(base_path, target_path);
+    normalized_path
+        .metadata()
+        .map_err(|e| (normalized_path.clone(), e))?;
+    builder
+        .append_path_with_name(&normalized_path, &target_path)
+        .map_err(|e| (normalized_path.clone(), e))?;
+    Ok(normalized_path)
+}
+
+fn on_error(e: anyhow::Error) -> anyhow::Error {
+    #[cfg(feature = "telemetry")]
+    sentry::integrations::anyhow::capture_anyhow(&e);
+
+    e
+}
+
+fn normalize_path(cwd: &Path, path: &Path) -> PathBuf {
+    let mut out = PathBuf::from(cwd);
+    let mut components = path.components();
+    if path.is_absolute() {
+        log::warn!(
+            "Interpreting absolute path {} as a relative path",
+            path.to_string_lossy()
+        );
+        components.next();
+    }
+    for comp in components {
+        out.push(comp);
+    }
+    out
+}
+
+/// Takes the package archive as a File and attempts to sign it using the active key
+/// returns the public key id used to sign it and the signature string itself
+pub fn sign_compressed_archive(
+    compressed_archive: &mut fs::File,
+) -> anyhow::Result<SignArchiveResult> {
+    let key_db = open_db()?;
+    let personal_key = if let Ok(v) = get_active_personal_key(&key_db) {
+        v
+    } else {
+        return Ok(SignArchiveResult::NoKeyRegistered);
+    };
+    let password = rpassword::prompt_password(format!(
+        "Please enter your password for the key pair {}:",
+        &personal_key.public_key_id
+    ))
+    .ok();
+    let private_key = if let Some(priv_key_location) = personal_key.private_key_location {
+        match minisign::SecretKey::from_file(&priv_key_location, password) {
+            Ok(priv_key_data) => priv_key_data,
+            Err(e) => {
+                log::error!(
+                    "Could not read private key from location {}: {}",
+                    priv_key_location,
+                    e
+                );
+                return Err(e.into());
+            }
+        }
+    } else {
+        // TODO: add more info about why this might have happened and what the user can do about it
+        log::warn!("Active key does not have a private key location registered with it!");
+        return Err(anyhow!("Cannot sign package, no private key"));
+    };
+    Ok(SignArchiveResult::Ok {
+        public_key_id: personal_key.public_key_id,
+        signature: (minisign::sign(
+            Some(&minisign::PublicKey::from_base64(
+                &personal_key.public_key_value,
+            )?),
+            &private_key,
+            compressed_archive,
+            None,
+            None,
+        )?
+        .to_string()),
+    })
```

Comment:
Let's split this into separate statements so it's easier to see where we're generating the signature.

```suggestion
    let public_key = minisign::PublicKey::from_base64(
        &personal_key.public_key_value,
    )?;
    let signature = minisign::sign(
            Some(&public_key),
            &private_key,
            compressed_archive,
            None,
            None,
    )?;
    Ok(SignArchiveResult::Ok {
        public_key_id: personal_key.public_key_id,
        signature: signature.to_string(),
    })
```